File: src/main/java/za/ac/sun/grapl/visitors/OpStackMethodVisitor.java
package za.ac.sun.grapl.visitors;

import org.objectweb.asm.Label;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.util.ASMifier;
import za.ac.sun.grapl.controllers.OpStackController;
import za.ac.sun.grapl.util.ASMParserUtil;

public class OpStackMethodVisitor extends MethodVisitor implements Opcodes {

    private final OpStackController controller;

    public OpStackMethodVisitor(
            final MethodVisitor mv,
            final OpStackController abstractController
    ) {
        super(ASM5, mv);
        this.controller = abstractController;
    }

    @Override
    public void visitCode() {
        super.visitCode();
        controller.initializeMethod();
    }

    @Override
    public void visitLabel(final Label label) {
        this.controller.pushNewLabel(label);
    }

    @Override
    public void visitInsn(final int opcode) {
        super.visitInsn(opcode);
        controller.pushConstInsnOperation(opcode);
    }

    @Override
    public void visitIntInsn(final int opcode, final int operand) {
        super.visitIntInsn(opcode, operand);
        controller.pushConstInsnOperation(opcode, operand);
    }

    @Override
    public void visitVarInsn(final int opcode, final int var) {
        super.visitVarInsn(opcode, var);
        final String operation = ASMifier.OPCODES[opcode];

        if (ASMParserUtil.isLoad(operation)) controller.pushVarInsnLoad(var, operation);
        else if (ASMParserUtil.isStore(operation)) controller.pushVarInsnStore(var, operation);
    }

    @Override
    public void visitJumpInsn(final int opcode, final Label label) {
        super.visitJumpInsn(opcode, label);
        final String jumpOp = ASMifier.OPCODES[opcode];

        if (ASMParserUtil.NULLARY_JUMPS.contains(jumpOp)) controller.pushNullaryJumps(label);
        else if (ASMParserUtil.UNARY_JUMPS.contains(jumpOp)) controller.pushUnaryJump(jumpOp, label);
        else if (ASMParserUtil.BINARY_JUMPS.contains(jumpOp)) controller.pushBinaryJump(jumpOp, label);
    }

    @Override
    public void visitLdcInsn(final Object val) {
        super.visitLdcInsn(val);
        controller.pushConstInsnOperation(val);
    }

    @Override
    public void visitIincInsn(final int var, final int increment) {
        super.visitIincInsn(var, increment);
        controller.pushVarInc(var, increment);
    }

}

File: src/main/java/za/ac/sun/grapl/visitors/OpStackMethodVisitor.kt
package za.ac.sun.grapl.visitors

import org.objectweb.asm.Label
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes
import org.objectweb.asm.util.ASMifier
import za.ac.sun.grapl.controllers.OpStackController
import za.ac.sun.grapl.util.ASMParserUtil
import za.ac.sun.grapl.util.ASMParserUtil.isLoad
import za.ac.sun.grapl.util.ASMParserUtil.isStore

open class OpStackMethodVisitor(
        mv: MethodVisitor?,
        private val controller: OpStackController
) : MethodVisitor(Opcodes.ASM5, mv), Opcodes {
    override fun visitCode() {
        super.visitCode()
        controller.initializeMethod()
    }

    override fun visitLabel(label: Label) {
        controller.pushNewLabel(label)
    }

    override fun visitInsn(opcode: Int) {
        super.visitInsn(opcode)
        controller.pushConstInsnOperation(opcode)
    }

    override fun visitIntInsn(opcode: Int, operand: Int) {
        super.visitIntInsn(opcode, operand)
        controller.pushConstInsnOperation(opcode, operand)
    }

    override fun visitVarInsn(opcode: Int, `var`: Int) {
        super.visitVarInsn(opcode, `var`)
        val operation = ASMifier.OPCODES[opcode]
        if (isLoad(operation)) controller.pushVarInsnLoad(`var`, operation) else if (isStore(operation)) controller.pushVarInsnStore(`var`, operation)
    }

    override fun visitJumpInsn(opcode: Int, label: Label) {
        super.visitJumpInsn(opcode, label)
        val jumpOp = ASMifier.OPCODES[opcode]
        when {
            ASMParserUtil.NULLARY_JUMPS.contains(jumpOp) -> controller.pushNullaryJumps(label)
            ASMParserUtil.UNARY_JUMPS.contains(jumpOp) -> controller.pushUnaryJump(jumpOp, label)
            ASMParserUtil.BINARY_JUMPS.contains(jumpOp) -> controller.pushBinaryJump(jumpOp, label)
        }
    }

    override fun visitLdcInsn(`val`: Any) {
        super.visitLdcInsn(`val`)
        controller.pushConstInsnOperation(`val`)
    }

    override fun visitIincInsn(`var`: Int, increment: Int) {
        super.visitIincInsn(`var`, increment)
        controller.pushVarInc(`var`, increment)
    }

}